File: sciurus17_examples/src/joint_values.cpp
// Copyright 2025 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://github.com/ros-planning/moveit2_tutorials/blob
// /a547cf49ff7d1fe16a93dfe020c6027bcb035b51/doc/move_group_interface
// /src/move_group_interface_tutorial.cpp

#include "angles/angles.h"
#include "moveit/move_group_interface/move_group_interface.hpp"
#include "rclcpp/rclcpp.hpp"

using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;

static const rclcpp::Logger LOGGER = rclcpp::get_logger("joint_values");

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto move_group_node = rclcpp::Node::make_shared("joint_values", node_options);
  // For current state monitor
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(move_group_node);
  std::thread([&executor]() {executor.spin();}).detach();

  MoveGroupInterface move_group_arm(move_group_node, "l_arm_group");
  // 駆動速度を調整する
  move_group_arm.setMaxVelocityScalingFactor(0.5);  // Set 0.0 ~ 1.0
  move_group_arm.setMaxAccelerationScalingFactor(0.5);  // Set 0.0 ~ 1.0

  // SRDFに定義されている"l_arm_init_pose"の姿勢にする
  move_group_arm.setNamedTarget("l_arm_init_pose");
  move_group_arm.move();

  // 現在角度をベースに、目標角度を設定する
  auto joint_values = move_group_arm.getCurrentJointValues();
  // 各ジョイントの角度を１つずつ変更する
  double target_joint_value = angles::from_degrees(-15.0);
  for (size_t i = 0; i < joint_values.size(); i++) {
    joint_values[i] = target_joint_value;
    move_group_arm.setJointValueTarget(joint_values);
    move_group_arm.move();
  }

  // 初期姿勢に戻す
  move_group_arm.setNamedTarget("l_arm_init_pose");
  move_group_arm.move();

  rclcpp::shutdown();
  return 0;
}

File: sciurus17_examples_py/sciurus17_examples_py/joint_values.py
# Copyright 2024 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from moveit.core.kinematic_constraints import construct_joint_constraint
from moveit.core.robot_state import RobotState
from moveit.planning import MoveItPy
from moveit.planning import PlanRequestParameters
import rclpy
from rclpy.logging import get_logger

from sciurus17_examples_py.utils import plan_and_execute


class JointValues:
    def __init__(self):
        # MoveItPyのインスタンスを生成し、planning componentを取得
        self.sciurus17 = MoveItPy(node_name='joint_values')
        self.logger = get_logger('joint_values')

        # アーム制御用 planning component
        self.arm = self.sciurus17.get_planning_component('l_arm_group')

        # ロボットモデルの取得（ジョイント目標値の設定に使用）
        self.robot_model = self.sciurus17.get_robot_model()

        # プランニングの設定（動作プランナーと速度・加速度スケール）
        self.arm_plan_params = PlanRequestParameters(
            self.sciurus17, 'ompl_rrtc_default'
        )
        self.arm_plan_params.max_velocity_scaling_factor = 0.1  # 0.0〜1.0の範囲で設定
        self.arm_plan_params.max_acceleration_scaling_factor = (
            0.1  # 0.0〜1.0の範囲で設定
        )

    def move_arm_to_named_pose(self, configuration_name):
        # SRDFに定義された姿勢名でアームを動かす
        self.arm.set_start_state_to_current_state()
        self.arm.set_goal_state(configuration_name=configuration_name)
        plan_and_execute(
            self.sciurus17,
            self.arm,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def move_arm_joint_values(self, joint_values_dict):
        # 各ジョイント角度[rad]を指定してアームを動かす
        # joint_values_dictはジョイント名をキー、角度[rad]を値とする辞書
        robot_state = RobotState(self.robot_model)
        robot_state.joint_positions = joint_values_dict

        joint_constraint = construct_joint_constraint(
            robot_state=robot_state,
            joint_model_group=self.robot_model.get_joint_model_group(
                'l_arm_group'
            ),
        )

        self.arm.set_start_state_to_current_state()
        self.arm.set_goal_state(motion_plan_constraints=[joint_constraint])
        plan_and_execute(
            self.sciurus17,
            self.arm,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def get_current_arm_joint_values(self):
        # アームの現在のジョイント角度を辞書形式で取得する
        current_state = self.arm.get_start_state()
        return current_state.get_joint_group_positions('l_arm_group')


def main(args=None):
    rclpy.init(args=args)

    controller = JointValues()

    joint_names = [
        'l_arm_joint1',
        'l_arm_joint2',
        'l_arm_joint3',
        'l_arm_joint4',
        'l_arm_joint5',
        'l_arm_joint6',
        'l_arm_joint7',
    ]
    target_joint_diff_value = math.radians(15.0)

    # l_arm_init_poseの姿勢にする
    controller.move_arm_to_named_pose('l_arm_init_pose')

    # 各関節角度を初期姿勢から順番に15[deg]ずつ動かす
    joint_values = controller.get_current_arm_joint_values()
    joint_values_dict = dict(zip(joint_names, joint_values))

    for joint_index, joint_name in enumerate(joint_names):
        if joint_values[joint_index] > 0.1:
            joint_values_dict[joint_name] -= target_joint_diff_value
        else:
            joint_values_dict[joint_name] += target_joint_diff_value
        controller.move_arm_joint_values(joint_values_dict)

    # l_arm_init_poseの姿勢に戻す
    controller.move_arm_to_named_pose('l_arm_init_pose')

    # 既知の不具合により終了時にエラーになるが問題ない。関連Issue:
    # https://github.com/moveit/moveit2/issues/2693
    rclpy.shutdown()


if __name__ == '__main__':
    main()
